### PR TITLE
Using DML library to offload read/write operations in pmem inside bluestore

### PR DESCRIPTION
**Pull request by optimistyzy**
The purpose is to offload memory/pmem operations
through hardware path in DML library

Signed-off-by: Ziye Yang <ziye.yang@intel.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 1]
```diff
@@ -377,3 +377,23 @@ settings to ensure that all IOs are issued through SPDK.::
...
 Otherwise, the current implementation will populate the SPDK map files with
 kernel file system symbols and will use the kernel driver to issue DB/WAL IO.
+
+
+DSA (Data Streaming Accelerator Usage)
+======================================
+
+If you want to use the DML library to drive DSA device for offloading
+read/write operations on Persist memory in Bluestore. You need to install
+DML and idxd-confgi library in your machine with SPR (Sapphire Rapids) CPU.
```
Comment (tchaikov) [inline doc/rados/configuration/bluestore-config-ref.rst:387]:
```suggestion
DML and idxd-confg library in your machine with SPR (Sapphire Rapids) CPU.
```

File: cmake/modules/Finddml.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+find_library(DML_HW_LIBRARIES NAMES dmlhl libdmlhl PATHS
+  /usr/local/
+  /usr/local/lib64
+  /usr/lib)
+
+list(APPEND DML_LIBRARIES ${DML_HW_LIBRARIES})
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(dml DEFAULT_MSG DML_LIBRARIES DML_INCLUDE_DIR)
+
+mark_as_advanced(DML_LIBRARIES DML_INCLUDE_DIR)
```
Comment (tchaikov) [inline cmake/modules/Finddml.cmake:28]:
```suggestion
 find_path(DMLHL_INCLUDE_DIR
   dml/dml.hpp
   PATHS
   /usr/include
   /usr/local/include)
find_library(DML_HL_LIBRARIES NAMES dmlhl libdmlhl PATHS
  /usr/local/
  /usr/local/lib64
  /usr/lib)

include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(dml DEFAULT_MSG
  DML_LIBRARIES
  DML_INCLUDE_DIR
  DML_HL_LIBRARIES
  DMLHL_INCLUDE_DIR)
mark_as_advanced(
  DML_LIBRARIES
  DML_INCLUDE_DIR
  DML_HL_LIBRARIES)

if(dml_FOUND)
  if(NOT (TARGET dml::dml))
    add_library(dml::dml UNKNOWN IMPORTED)
    set_target_properties(dml::dml PROPERTIES
      INTERFACE_INCLUDE_DIRECTORIES "${DML_INCLUDE_DIR}"
      IMPORTED_LINK_INTERFACE_LANGUAGES "C"
      IMPORTED_LOCATION "${DML_LIBRARIES}")
  endif()
  if(NOT (TARGET dml::dmlhl))
    add_library(dml:: dmlhl UNKNOWN IMPORTED)
    set_target_properties(dml::dmlhl PROPERTIES
      INTERFACE_INCLUDE_DIRECTORIES "${DMLHL_INCLUDE_DIR}"
      INTERFACE_LINK_LIBRARIES ${CMAKE_DL_LIBS}
      INTERFACE_COMPILE_FEATURES cxx_std_17
      IMPORTED_LINK_INTERFACE_LANGUAGES "CXX"
      IMPORTED_LOCATION "${DMLHL_LIBRARIES}")
  endif()
```

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 1 above)
Comment (optimistyzy) [reply doc/rados/configuration/bluestore-config-ref.rst:387]: done

File: cmake/modules/Finddml.cmake [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+if(DML_FOUND)
+  if(NOT (TARGET dml::dml))
+    add_library(dml::dml UNKNOWN IMPORTED)
+    set_target_properties(dml::dml PROPERTIES
+      INTERFACE_INCLUDE_DIRECTORIES "${DML_INCLUDE_DIR}"
+      IMPORTED_LINK_INTERFACE_LANGUAGES "C"
+      IMPORTED_LOCATION "${DML_LIBRARIES}")
+  endif()
+
+ if(NOT (TARGET dml::dmlhl))
```
Comment (tchaikov) [inline cmake/modules/Finddml.cmake:48]: wrong indent.

File: cmake/modules/Finddml.cmake [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+# Find the dml and dmlhl libraries and includes
+#
+# DML_INCLUDE_DIR - where to find dml.hpp etc.
+# DML_LIBRARIES - List of libraries when using dml.
+# DML_HL_LIBRARIES - List of libraries when using dmlhl.
+# DML_FOUND - True if DML found.
+
+
+find_path(DML_INCLUDE_DIR
+   dml/dml.hpp
```
Comment (tchaikov) [inline cmake/modules/Finddml.cmake:11]: nit, inconsistent indent.

Comment (optimistyzy) [reply cmake/modules/Finddml.cmake:11]: ack

File: cmake/modules/Finddml.cmake (hunk 3 above)
Comment (optimistyzy) [reply cmake/modules/Finddml.cmake]: ack

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 5]
```diff
@@ -452,3 +452,21 @@ of ``/sys/block/<drive>/queue/optimal_io_size``.
...
+DSA (Data Streaming Accelerator Usage)
+======================================
+
+If you want to use the DML library to drive DSA device for offloading
+read/write operations on Persist memory in Bluestore. You need to install
+DML and idxd-config library in your machine with SPR (Sapphire Rapids) CPU.
+Refer following site for more details.
+
+.. __: https://github.com/intel/DML
+.. __: https://github.com/intel/idxd-config
```
Comment (tchaikov) [inline doc/rados/configuration/bluestore-config-ref.rst:464]: @optimistyzy  these hyperlinks lead to following warning:

```
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/rados/configuration/bluestore-config-ref.rst:: ERROR: Anonymous hyperlink mismatch: 1 references but 3 targets.
```

and the generated HTML does not look right. see https://ceph--44230.org.readthedocs.build/en/44230/rados/configuration/bluestore-config-ref/#dsa-data-streaming-accelerator-usage

could you please use the following notation instead? something like
```rst
- `DML <https://github.com/intel/DML>`__
- `libaccel-config <https://github.com/intel/idxd-config/>`__
```

File: src/blk/CMakeLists.txt [hunk 6]
```diff
@@ -44,6 +44,10 @@ if(WITH_ZBD)
 endif()
 
 if(WITH_BLUESTORE_PMEM)
+  if(HAVE_LIBDML)
+    target_link_libraries(blk PRIVATE dml::dml dml::dmlhl)
+  endif(HAVE_LIBDML)
```
Comment (tchaikov) [inline src/blk/CMakeLists.txt:49]: i'd suggest remove `HAVE_LIBDML` in the `endif()`.

File: src/blk/pmem/PMEMDevice.cc [hunk 7]
```diff
@@ -28,6 +28,12 @@
 #include "common/debug.h"
 #include "common/blkdev.h"
 
+#if defined(HAVE_LIBDML)
+#include <dml/dml.hpp>
+using execution_path = dml::automatic;
+using namespace std;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:34]: why is this `using namespace std` necessary?

File: src/blk/pmem/PMEMDevice.cc [hunk 8]
```diff
@@ -219,7 +225,14 @@ int PMEMDevice::write(uint64_t off, bufferlist& bl, bool buffered, int write_hin
   while (len) {
     const char *data;
     uint32_t l = p.get_ptr_and_advance(len, &data);
+
+#if defined(HAVE_LIBDML)
+  // Take care of the persistency issue
+  auto result = dml::execute<execution_path>(dml::mem_move, dml::make_view(data, l), dml::make_view(addr + off1, l));
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:231]: wrong indent.

File: src/blk/CMakeLists.txt (hunk 6 above)
Comment (optimistyzy) [reply src/blk/CMakeLists.txt:49]: ack

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 5 above)
Comment (optimistyzy) [reply doc/rados/configuration/bluestore-config-ref.rst:464]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 8 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:231]: ack

File: src/blk/CMakeLists.txt (hunk 6 above)
Comment (optimistyzy) [reply src/blk/CMakeLists.txt:49]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 7 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc]: Let me remove it.

**Comment by optimistyzy** [comment]
@tchaikov Doc related issue is fixed.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 9]
```diff
@@ -452,3 +452,21 @@ of ``/sys/block/<drive>/queue/optimal_io_size``.
...
+DSA (Data Streaming Accelerator Usage)
+======================================
+
+If you want to use the DML library to drive DSA device for offloading
+read/write operations on Persist memory in Bluestore. You need to install
+`DML`_ and `idxd-config`_ library in your machine with SPR (Sapphire Rapids) CPU.
+Refer following site for more details.
+
+.. _DML: https://github.com/intel/DML
+.. _idxd-config: https://github.com/intel/idxd-config`
```
Comment (tchaikov) [inline doc/rados/configuration/bluestore-config-ref.rst:464]: remove the trailing "`". the link in https://ceph--44230.org.readthedocs.build/en/44230/rados/configuration/bluestore-config-ref/ brings me to https://github.com/intel/idxd-config%60 which is a 404.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 10]
```diff
@@ -452,3 +452,21 @@ of ``/sys/block/<drive>/queue/optimal_io_size``.
...
+
+After installing the DML software, you need to configure the shared
+work queues (WQs) with the following WQ configuration example via accel-config tool::
```
Comment (tchaikov) [inline doc/rados/configuration/bluestore-config-ref.rst:467]: nit, might want use 

```rst
.. code-block:: console
```

for better readability.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 9 above)
Comment (optimistyzy) [reply doc/rados/configuration/bluestore-config-ref.rst:464]: ack

**Review by tchaikov — DISMISSED** [review]

File: cmake/modules/Buildpmem.cmake [hunk 11]
```diff
@@ -48,6 +49,9 @@ function(build_pmem)
   set_target_properties(pmem::pmem PROPERTIES
     INTERFACE_INCLUDE_DIRECTORIES ${PMDK_INCLUDE}
     IMPORTED_LOCATION "${PMDK_LIB}/libpmem.a"
+    IMPORTED_LOCATION "${PMDK_LIB}/libpmem2.a"
+    IMPORTED_LOCATION "${PMDK_LIB}/libpmemcommon.a"
+    IMPORTED_LOCATION "${PMDK_LIB}/libpmemcore.a"
```
Comment (tchaikov) [inline cmake/modules/Buildpmem.cmake:54]: @optimistyzy  this fails the build. 

```
[100%] Linking CXX executable ../../../bin/unittest_librbd
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_memmove'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_deep_flush'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_msync'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_persist'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_flush'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_memset'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_drain'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_is_pmem'
../../../lib/libceph_librbd_pwl_cache.so.1.0.0: undefined reference to `pmem_memcpy'
collect2: error: ld returned 1 exit status
make[2]: *** [src/test/librbd/CMakeFiles/unittest_librbd.dir/build.make:1652: bin/unittest_librbd] Error 1
make[1]: *** [CMakeFiles/Makefile2:16137: src/test/librbd/CMakeFiles/unittest_librbd.dir/all] Error 2
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/62138//consoleFull

it seems that we can have only a single `IMPORTED_LOCATION` for each target. see https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED_LOCATION.html.

an alternative is to have multiple targets for each of these pmdk libraries, something like
```cmake
foreach(c pmem pmem2 pmemcommon pmemcore pmemobj)
  add_library(pmem::$c STATIC IMPORTED GLOBAL)
  add_dependencies(pmem::$c pmdk_ext)
  set_target_properties(pmem::$c PROPERTIES
    INTERFACE_INCLUDE_DIRECTORIES ${PMDK_INCLUDE}
    IMPORTED_LOCATION "${PMDK_LIB}/lib${c}.a")
endforeach()
set_property(TARGET pmem::pmemobj
  PROPERTY INTERFACE_LINK_LIBRARIES "pmem::pmem;${CMAKE_THREAD_LIBS_INIT}")
# add more inter-library dependencies here
```

and (only) link against the used pmem libraries.

**Comment by optimistyzy** [comment]
Just checked, I think that I add a wrong commit inside.

File: cmake/modules/Buildpmem.cmake (hunk 11 above)
Comment (optimistyzy) [reply cmake/modules/Buildpmem.cmake]: This is a local test version, which should not be included in this PR.

Comment (tchaikov) [reply cmake/modules/Buildpmem.cmake]: fair enough. will retest this PR.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
pmdk/pmem support is disabled by default, so i am going ti merge this changeset without running it against the rados qa suites.

**Pull request merged**